async.c

```diff
@@ -653,25 +679,109 @@ static int __redisAsyncCommand(redisAsyncContext *ac, redisCallbackFn *fn, void
         /* (P)UNSUBSCRIBE does not have its own response: every channel or
          * pattern that is unsubscribed will receive a message. This means we
          * should not append a callback function for this command. */
-     } else if(strncasecmp(cstr,"monitor\r\n",9) == 0) {
-         /* Set monitor flag and push callback */
-         c->flags |= REDIS_MONITORING;
-         __redisPushCallback(&ac->replies,&cb);
+    } else if(strncasecmp(cstr,"monitor\r\n",9) == 0) {
+        /* Set monitor flag and push callback */
+        c->flags |= REDIS_MONITORING;
+        SAFE_PUSHCALLBACK(replies)
     } else {
-        if (c->flags & REDIS_SUBSCRIBED)
+        if (c->flags & REDIS_SUBSCRIBED) {
             /* This will likely result in an error reply, but it needs to be
              * received and passed to the callback. */
-            __redisPushCallback(&ac->sub.invalid,&cb);
-        else
-            __redisPushCallback(&ac->replies,&cb);
+            SAFE_PUSHCALLBACK(sub.invalid)
+        } else {
+            SAFE_PUSHCALLBACK(replies)
+        }
     }
+#undef SAFE_PUSHCALLBACK
 
-    __redisAppendCommand(c,cmd,len);
+    return REDIS_OK;
 
-    /* Always schedule a write when the write buffer is non-empty */
-    _EL_ADD_WRITE(ac);
+oom_error:
+    /* Rollback the obuf [size still allocated] */
+    clen = sdslen(c->obuf) - len;
```

Comment:
If reacting to an OOM, `sdsRemoveFreeSpace()` is helpful too. https://github.com/redis/hiredis/blob/master/sds.h#L257

Reply:
Added call to sdsRemoveFreeSpace() in hopes it will help the users